Add HNS Load Balancer Healthchecks for ExternalTrafficPolicy: Local

anfernee opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it:**
In GCE, the current `externalTrafficPolicy: Local` logic does not work because the nodes that run the pods do not setup load balancer ports. This means that the GCLB does not understand which nodes are serving the pods that can accept traffic. Since all report unhealthy it'll direct traffic to any node. This PR configures the health check ports so that GCLB knows which nodes can handle the traffic.

See #62046 for details.
This work is based on #96998 

**Which issue(s) this PR fixes:**

Fixes #62046

**Special notes for your reviewer:**

Does this PR introduce a user-facing change?:

`externalTrafficPolicy: local` works on Windows hosts with DSR support.

**Release Note**
```release-note
Add 2 new options for kube-proxy running in winkernel mode. 
`--forward-healthcheck-vip`, if specified as true, health check traffic whose destination is service VIP will be forwarded to kube-proxy's healthcheck service. `--root-hnsendpoint-name` specifies the name of the hns endpoint for the root network namespace.
This option enables the pass-through load balancers like Google's GCLB to correctly health check the backend services. Without this change, the health check packets is dropped, and Windows node will be considered to be unhealthy by those load balancers.
```

jeremyje commented:
/sig windows
/priority important-soon

anfernee commented:
/retest

anfernee commented:
@madhanrm @elweb9858 PTAL.

jeremyje commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1270,6 +1283,24 @@ func (proxier *Proxier) syncProxyRules() {
+			if gceGatewayEndpoint != nil {
+				hnsHealthCheckLoadBalancer, err := hns.getLoadBalancer(
+					[]endpointsInfo{*gceGatewayEndpoint},
+					loadBalancerFlags{isDSR: svcInfo.preserveDIP || proxier.isDSR, useMUX: svcInfo.preserveDIP, preserveDIP: svcInfo.preserveDIP},
```
Should the flags here be synced with the flags above?

anfernee replied:
Actually I am not sure what this annotation means here: `preserveDIP := service.Annotations["preserve-destination"] == "true"`, they are copied over.

The rest is not necessarily synced, because the only targeting client is GCLB health check, the requirement for the service itself doesn't apply to health check.

anfernee commented:
friendly ping.. @andrewsykim @ksubrmnn

anfernee committed:
Fix health check from Google's Load Balancer

This change adds 2 options for windows:
--forward-healthcheck-vip: If true forward service VIP for health check
port
--root-hnsendpoint-name: The name of the hns endpoint name for root
namespace attached to l2bridge, default is cbr0

When --forward-healthcheck-vip is set as true and winkernel is used,
kube-proxy will add an hns load balancer to forward health check request
that was sent to lb_vip:healthcheck_port to the node_ip:healthcheck_port.
Without this forwarding, the health check from google load balancer will
fail, and it will stop forwarding traffic to the windows node.

This change fixes the following 2 cases for service:
- `externalTrafficPolicy: Cluster` (default option): healthcheck_port is
10256 for all services. Without this fix, all traffic won't be directly
forwarded to windows node. It will always go through a linux node and
get forwarded to windows from there.
- `externalTrafficPolicy: Local`: different healthcheck_port for each
service that is configured as local. Without this fix, this feature
won't work on windows node at all. This feature preserves client ip
that tries to connect to their application running in windows pod.

Change-Id: If4513e72900101ef70d86b91155e56a1f8c79719

daschott commented:
@elweb9858 @Keith-Mange PTAL

elweb9858 commented:
@sbangari could you take a look at this?

anfernee commented:
@ibabou one more try to have this merged.

ibabou commented:
/retest

marosset commented:
@anfernee / @sbangari is this ready for review again?

anfernee commented:
> @anfernee / @sbangari is this ready for review again?

Yes! I added 2 options to guard the behavior as @thockin suggested. Appreciated it if you can take a look.

thockin commented:
Is this shooting for 23?

anfernee commented:
> Is this shooting for 23?

Yes, it would be nice to include in.

jyotimahapatra commented:
Hi 👋 I'm the bug triage shadow. We're in code freeze for 1.23. Just wanted to check if we can still get this PR merged in time for release(7/12)

jyotimahapatra commented:
Hi 👋 I'm the bug triage shadow. We're in code freeze for 1.23 milestone. I wanted to check if we're targeting to get this PR merged in time for 1.23 release(7/12)

marosset commented:
/milestone v1.24

aojea commented on cmd/kube-proxy/app/server_windows.go:
```diff
@@ -97,8 +99,11 @@ func newProxyServer(config *proxyconfigapi.KubeProxyConfiguration, cleanupAndExi
+	var healthzPort int
 	if len(config.HealthzBindAddress) > 0 {
 		healthzServer = healthcheck.NewProxierHealthServer(config.HealthzBindAddress, 2*config.IPTables.SyncPeriod.Duration, recorder, nodeRef)
+		_, port, _ := net.SplitHostPort(config.HealthzBindAddress)
```
this seems is ok since is already validated in https://github.com/kubernetes/kubernetes/blob/a53e2eaeaba064309dceca2dc27f3ac09c6375b0/pkg/proxy/apis/config/validation/validation.go

aojea commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -1302,6 +1321,28 @@ func (proxier *Proxier) syncProxyRules() {
 			klog.V(3).InfoS("Hns LoadBalancer resource created for loadBalancer Ingress resources", "lbIngressIP", lbIngressIP)
+
+			if proxier.forwardHealthCheckVip && gatewayHnsendpoint != nil {
+				nodeport := proxier.healthzPort
```
but this is not correct, this is about the healthz of the apiserver, not about if the service has local endpoints ... but to be honest I don't fully understand the problem, what is the TL; DR, of why the NewServiceHealthServer doesn't work in windows? that should fix the problem, no?
```
	serviceHealthServer := healthcheck.NewServiceHealthServer(hostname, recorder, []string{} /* windows listen to all node addresses */)
```

anfernee replied:
The problem is it's not the node address. The google load balancer works this way:

```
User --> [clientip:12345 to lb_vip:80] -> Google LB -> [clientip:12345 to lb_vip:80] -> Node -> [clientip:12345 to pod_ip:80] -> Pod 
```
This works right now in kubeproxy's winkernel mode, because the rules are already in place.

What doesn't work is the health check for normal service:
```
Google LB -> [lb:12345 to lb_vip:10256] -> Node -> X(dropped)
```
and for `externalTrafficPolicy: local` service
```
Google LB -> [lb:12345 to lb_vip:<per-service-hc-port>] -> Node -> X(dropped)
```
They are dropped because the destination is not among any node IPs.

This basically adds hns rules to change request to node IP.

aojea replied:
yeah, but why it uses the kube-proxy healthz port , it should use the healthzcheck nodeport, no?

aojea replied:
the part I'm lost is about when it has to use healthz port, but I'm not familiar with the internals of GCE LBs, maybe we should get a reviewer more familiar with this :sweat_smile:

anfernee replied:
> yeah, but why it uses the kube-proxy healthz port , it should use the healthzcheck nodeport, no?

`proxier.healthzPort` is basically 10256 if it's not explicitly configured: 

https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/cmd/kube-proxy/app/server.go#L212-L220

https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/cmd/kube-proxy/app/server.go#L226

https://github.com/kubernetes/kubernetes/blob/1bd00776b5d78828a065b5c21e7003accc308a06/pkg/cluster/ports/ports.go#L36

this is basically the healthzcheck nodeport. For `externalTrafficPolicy: local`, the port is different for every service. 

> the part I'm lost is about when it has to use healthz port, but I'm not familiar with the internals of GCE LBs, maybe we should get a reviewer more familiar with this 😅

I understand it's really not straight-forward.. I'll try to get someone from GKE service team to review as well :) @robscott @bowei can you guys help take a look?

robscott replied:
Just catching up on this thread. I _think_ this makes sense. Although we don't need to do this in other proxier implementations, it makes sense that we'd need to this lb_vip -> node ip translation here to get health checks working. I'm less familiar with how the primary/default health check needs to work here. Regardless of OS, that seems to be handled exclusively at a higher level, with no extra logic added in either iptables or winkernel proxier implementations:

https://github.com/kubernetes/kubernetes/blob/481459d12dc82ab88e413886e2130c2a5e4a8ec4/cmd/kube-proxy/app/server_others.go#L153
https://github.com/kubernetes/kubernetes/blob/481459d12dc82ab88e413886e2130c2a5e4a8ec4/cmd/kube-proxy/app/server_windows.go#L101

I'm not familiar enough with how this is setup to know what's necessary, but is it possible that this works without any changes for the default health check? The original bug as reported by the PR seems to indicate the problem is only with `ExternalTrafficPolicy: Local`.

anfernee replied:
at higher level, `healthcheck.NewProxierHealthServer()` is called to create http server that listens on the specific bind address for health check purpose. Because it listens on `0.0.0.0`, on Linux it will accept the connection regardless of destination. Unfortunately for Windows, the additional forward needs to be done to forward VIP to node-ip so the health check server can receive traffic.

robscott commented on cmd/kube-proxy/app/init_windows.go:
```diff
@@ -41,4 +41,6 @@ func (o *Options) addOSFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&o.config.Winkernel.SourceVip, "source-vip", o.config.Winkernel.SourceVip, "The IP address of the source VIP for non-DSR.")
 	fs.StringVar(&o.config.Winkernel.NetworkName, "network-name", o.config.Winkernel.NetworkName, "The name of the cluster network.")
 	fs.BoolVar(&o.config.Winkernel.EnableDSR, "enable-dsr", o.config.Winkernel.EnableDSR, "If true make kube-proxy apply DSR policies for service VIP")
+	fs.StringVar(&o.config.Winkernel.RootHnsendpointName, "root-hnsendpoint-name", "cbr0", "The name of the hns endpoint name for root namespace attached to l2bridge")
```
Nit: I think `HnsEndpoint` is a better way to represent this name throughout. That would also match the capitalization above in k8s-node-setup.psm1.

robscott commented on pkg/proxy/winkernel/hnsV1.go:
```diff
@@ -108,6 +109,22 @@ func (hns hnsV1) getEndpointByIpAddress(ip string, networkName string) (*endpoin
+	}
+	return &endpointsInfo{
+		ip:         hnsendpoint.IPAddress.String(),
+		isLocal:    !hnsendpoint.IsRemoteEndpoint, //TODO: Change isLocal to isRemote
```
Nit: This TODO (and the other similar one) may fit a bit better on the type definition.

robscott reviewed (commented):
Thanks for the work on this @anfernee!

robscott commented:
Thanks @anfernee!

/remove-lifecycle stale
/triage accepted
/lgtm

robscott commented:
/retest

aojea commented:
/approve
/hold
for windows folks to take a look and unhold when they are happy
/assign @jsturtevant @jayunit100

jayunit100 commented:
this is tricky to reason about....   the urgency on merging this patch ? we could merge it now, watch it closely, and iterate where needed OR, we could ... 
- along with this patch,  add a `sig-windows` tagged test that ensures that `externalTrafficPolicy=local` when implemented, always sends out correct health checks?  i think it would be doable ? 
- otherwise, if you file an issue in https://github.com/K8sbykeshed/k8s-service-lb-validator/, our new service lb tests we can try to create this test for you .

jayunit100 commented:
we have  service loadbalancing tests in the e2es triggered above: 

```
- Kubernetes e2e suite: [sig-windows] Services should be able to create a functioning NodePort service for Windows
- Kubernetes e2e suite: [sig-network] Services should be able to change the type from ExternalName to ClusterIP [Conformance]
- Services should serve multiport endpoints from pods
```

jayunit100 commented:
since the tests are passing, i think we should just merge and watch it, iterate if necessar

/lgtm
/approve

anfernee commented:
> imo we could merge this but , quick question... this is tricky to reason about....
> 
> whats the urgency on merging this patch ? we could...
> 
> 1. merge it now, watch it closely, and iterate where needed OR, we could ...
> 2. do a test coverage based approach
> 
> * along with this patch,  add a `sig-windows` tagged test that ensures that `externalTrafficPolicy=local` when implemented, always sends out correct health checks?  i think it would be doable ?
> * otherwise, if you file an issue in https://github.com/K8sbykeshed/k8s-service-lb-validator/, our new service lb tests we can try to create this test for you .
> 
> side note - I THINK we SKIP the ESIPP tests for IPVS... so it seems like the overall impl externalTrafficPolicy might not be uniform in kube proxy
> 
> anyways lmk what you think is best Yongkun Anfernee Gui ! if antonio and rob are on board, then merge+iterate seems like an ok idea to me.... i dont fully understand the kube proxy health check parts so ill have to learn more about it today as i read this.

since default behavior is unchanged, this should be safe to merge. In terms of whether this fix pass-through load balancer, we need to turn on those 2 options and find a pass-through lb (Google's GCLB) to check. Since this is very provider specific. Where do you recommend the test should be living in?

marosset commented:
/assign @thockin
For final review (or re-assignment)

thockin commented:
/approve

anfernee commented:
Thanks everyone for all the help for this PR. Finally almost there. Special thanks for the original author of this PR @jeremyje

anfernee commented:
When can we merge this PR?

jayunit100 commented:
/lgtm
/approve

jayunit100 commented:
Agree it's fine -> Let's merge and keep an eye out

jayunit100 commented:
Double checking that you got the earlier note: we should make sure our externalTrafficPolicy.... tests are covering stuff like this somehow, eventually, on the sig-win side

anfernee commented:
> Double checking that you got the earlier note: we should make sure our externalTrafficPolicy.... tests are covering stuff like this somehow, eventually, on the sig-win side

I am not exactly familiar with the test setup but I'll check it out. If the test runs on GCE, then we can enable the option there. Existing e2e service test should be able to catch it.

jsturtevant commented:
/hold cancel

anfernee commented:
/retest

anfernee commented:
Fixed `pkg/generated/openapi/zz_generated.openapi.go`. Pushed again.

anfernee commented:
This PR requires LGTM again @jayunit100 @jsturtevant

jayunit100 commented:
/lgtm

anfernee commented:
/retest
